### DISCUSSION TITLE
HM 600 Limit via MQTT setzen

### DISCUSSION THREAD

**Discussion by Mogdar-M** [tbnobody/OpenDTU#874] [Q&A] [answered]
Ich bin von Ahoy auf openDTU on battery gewechselt und der HM und der Victron sind sauber verbunden.
Daten melde ich via MQTT an den IO Broker.

Via Ahoy hatte ich einen Script, der den Wechselrichter abhängig von der Tageszeit limitiert.
Letzte Einstellung via Ahoy war 72%.

Nun finde ich keinen Weg, dieses Limit wieder zu ändern.
Weder in der Web GUI kann ich das ändern noch ist der Datenpunkt via MQTT erreichbar

Ich habe die Diskussion zur MQTT Struktur gesehen.
Es fehlt bei mir der Bereich cmd
cmd

- limit_persistent_relative
- limit_persistent_absolute
- limit_nonpersistent_relative
- limit_nonpersistent_absolute
- power
- restart

Und damit kann ich auch den HM nicht steuern.

Status zeigt mir den aktuellen Wert an.

Aber 
setState("mqtt.0.openDTU.11418xxxxxx./cmd/limit_persistent_relative"/openDTU/11418xxxxxx//cmd/limit_persistent_relative/,71);

ändert nichts, das eben die Datenfelder nicht da sind.

Wie bekomme ich das Limit eingestellt?

**Comment by DPO99** [comment]
> Ich bin von Ahoy auf openDTU on battery gewechselt und der HM und der Victron sind sauber verbunden. Daten melde ich via MQTT an den IO Broker.
> 
> Via Ahoy hatte ich einen Script, der den Wechselrichter abhängig von der Tageszeit limitiert. Letzte Einstellung via Ahoy war 72%.
> 
> Nun finde ich keinen Weg, dieses Limit wieder zu ändern. Weder in der Web GUI kann ich das ändern noch ist der Datenpunkt via MQTT erreichbar
>


Das Limit kannst du z.B. (einmalig) direkt in der DTU ändern. Warum sollte das denn da nicht gehen? Wird der Befehl als fehlgeschlagen übermittelt? Das Limit sollte sich auch nach einem Poll in der DTU wiederfinden.

![image](https://user-images.githubusercontent.com/125746530/235694302-35ec5ed0-38db-4e07-bff1-68f7ca5d3b25.png)



> Ich habe die Diskussion zur MQTT Struktur gesehen. Es fehlt bei mir der Bereich cmd cmd
> 
> * limit_persistent_relative
> * limit_persistent_absolute
> * limit_nonpersistent_relative
> * limit_nonpersistent_absolute
> * power
> * restart


Die sind per Default nicht da. Entweder legst du diese händisch an oder du nimmst z.B. den MQTT-Explorer und schiebst diese Werte einmal zum WR, danach sind diese auch im MQTT Broker da.

> Und damit kann ich auch den HM nicht steuern.
> 
> Status zeigt mir den aktuellen Wert an.
> 
> Aber setState("mqtt.0.openDTU.11418xxxxxx./cmd/limit_persistent_relative"/openDTU/11418xxxxxx//cmd/limit_persistent_relative/,71);
> 
> ändert nichts, das eben die Datenfelder nicht da sind.
> 
> Wie bekomme ich das Limit eingestellt?

Ich denke mit meinen Hinweisen oben, sollte das dann funktionieren. 
Ein Tipp noch: Ändere Dein Skript auf "nonpersistant". Der "persistant" Befehl belastet den Speicher, da der Wert ja fix im System (jedes mal) hinterlegt werden muss. "nonpersistant" wird nur im ram gespeichert.

Und zu guter letzt bist du hier eigentlich nicht ganz richtig, da hier nur die "originale" DTU zu finden ist. Für die OnBattery Version gibt es einen eigenen [Github](https://github.com/helgeerbe/OpenDTU-OnBattery).

**Reply by Mogdar-M** [reply]
Erst einmal Danke für die Antwort.

Natürlich war ich erst bei openDTU on battery, aber da es ja kein Thema der Erweiterung ist, gab es da auch keine Lösung.

Ich habe jetzt die Datenpunkte angelegt und exakt so benannt.
![image](https://user-images.githubusercontent.com/129499239/235703317-84d365a5-955e-4f29-9e70-907dc05a5d5e.png)

Ich war zu blind, die Einstellung via openDTU habe ich nun gefunden und die stellt auch den Wert ein.
Der wird dann auch unter Status angezeigt.

Und meine Scripts schieben jetzt Daten in die Felder.
Nur ändert sich damit nicht die Leistung des WR.

**Comment by DPO99** [comment, marked as answer]
Also ich kann Dir nur sagen wie ich das unter dem IoBroker mache und das läuft so auch seit einem halben Jahr stabil. Grundsätzlich sollte MQTT gleich MQTT sein aber fakt ist ja es geht nicht.

Ich habe die Instanz MQTT.0 also Broker eingerichtet. Diese Settings verwende ich:
![image](https://user-images.githubusercontent.com/125746530/235744665-d8c5c917-2767-4de3-a4a4-7bc0de406d9c.png)

![image](https://user-images.githubusercontent.com/125746530/235744718-021a0a8d-9aa0-4441-9179-d8e344fc6551.png)


Zusätzlich habe ich eine MQTT.1 Instanz als Client (verbunden mit MQTT.0) angelegt. Diese Settings verwende ich:
![image](https://user-images.githubusercontent.com/125746530/235744848-e039bda1-2b08-465a-838a-e43c7342ef1e.png)

![image](https://user-images.githubusercontent.com/125746530/235744907-c83d3d8b-74d9-417d-a9d1-4de44a4304d8.png)

Die OpenDTU habe ich dann mit den Anmeldeinformationen des MQTT.0 Brokers verbunden.

Ich steuere alle Objekte im Objektbaum immer in der MQTT.1 Instanz (cmd/Power + /cmd/limit nonpersistent absolute). Wichtig ist, dass man im Blockly den Befehl "steuere" verwendet.

Ich habe das Gefühl, dass es im IoBroker mit dem MQTT Adapter "Probleme" mit dem Retain Flag gibt. Manche MQTT Topics muss man ja mit und manche ohne senden. Ich war bisher zu doof das über Blockly einzustellen (auch mit den Settings im MQTT Adapter war ich nicht erfolgreich) also mache ich das eben über eine weitere MQTT Instanz. Es gibt sicher pfiffigere Lösungen aber meine ist zum Glück nicht allzu umständlich und funktioniert.

**Reply by Mogdar-M** [reply]
nochmal Danke für dein Feedback.

Ich habe jetzt auch mqtt 1 eingerichtet und dort dann die Datenpunkte angelegt.
Das Ergebnis bleibt das gleiche, der Wert vom Script wird in das Datenfeld geschrieben, doch es gibt keine Übertragung an den HM.

Unter Ahoy war es ja unter ctrl/limit/0
Also habe ich auch das mal angelegt und auch das sendet keine Daten an den HM
Da es unter Ahoy funktioniert hat vermute ich, dass ich irgendwas an dem Datenpunkt falsch habe.
Denn irgendwo im Code muss doch stehen, welcher Wert an den HM gesendet wird.
Wenn ich also statt/ dann einen "." habe oder ein Leerzeichen zu viel,
den Datenpunkt mit falschen Parameter angelegt habe etc geht es nicht.

Ursache bleibt ja, dass bei der MQTT Kommunikation diese Datenpunkte nicht angelegt werden.
Und offensichtlich bekomme ich sie auch nicht manuell eingerichtet.

**Reply by DPO99** [reply]
Das sollte aber nicht sein. Sofern der MQTT.1 ein Client/Subscriber ist, sollte er die CMD Punkte automatisch in der MQTT.1 Struktur anzeigen, welche Du im Broker MQTT.0 angelegt hast. Oder hattest Du diese wieder gelöscht? Es ist ja nur ein Client, der liest ja das aus dem Broker.

In den MQTT Topics ist ja genau beschrieben, wie der Punkt anzulegen ist. Und wenn du dort einen Wert änderst, wird dieser auch so wie im Broker angelegt an den HM übergeben. Zeig mal bitte Deine MQTT einstellungen für die 0 und die 1 und den cmd Baum für die 0 und die 1 und einen Screenshot von Deinem Blockly Skript. Iwo muss da was falsch sein.

**Reply by Mogdar-M** [reply]
ja, ist als Subcriber eingestellt
![image](https://user-images.githubusercontent.com/129499239/235871750-b3070c71-e74e-4cb6-a921-6ad743acd878.png)

Datenpunkte als Number konfiguriert, auch read und write
{
  "common": {
    "name": "limit_nonpersistent_absolute",
    "desc": "Manuell erzeugt",
    "role": "state",
    "type": "number",
    "read": true,
    "write": true,
    "def": 0
  },
  "type": "state",
  
  Einstellungen von MQTT1
  
![image](https://user-images.githubusercontent.com/129499239/235872897-01e2b1ce-bcc7-455a-addd-ec7639bf2282.png)
MQTT 0 genauso, nur als Server/Broker

und so sieht das in IO Broker aus

![image](https://user-images.githubusercontent.com/129499239/235871621-c5fe8628-f667-4215-98d6-7c484116cf0b.png)

**Reply by Mogdar-M** [reply]
![image](https://user-images.githubusercontent.com/129499239/235874212-ba6d3fb9-61f6-43d3-b99c-47e93e0b326d.png)
und so sieht der Blocky aus

zum Testen nutze ich dann diesen:
  setState("mqtt.0.openDTU.1141xxxxx.cmd.limit_persistent_relative",70, true)

**Comment by DPO99** [comment]
Werden in der MQTT.0 Instanz auch alle Werte der Wechselrichter angezeigt? Also Leistung, Spannung, einzelne Panele, etc.? Diese Werte sollten eigentlich auch in der MQTT.1 Instanz angezeigt werden, die sieht aber recht leer aus.

Und ich würde Dir empfehlen den Haken in der MQTT.0 und MQTT.1 Instanz bei "Eigene States beim verbinden publizieren" rauszunehmen, weil er sonst bei einem Restart von DTU oder Iobroker alles auf eimal rüber schiebt und das packt die openDTU nicht.

Bei mir sieht der Befehl auch anders aus:
![image](https://user-images.githubusercontent.com/125746530/235876656-9d24f623-f21e-4fdd-9245-20eb02da121b.png)

**EDIT**

**Reply by Mogdar-M** [reply]
alle ist ein großes Wort, aber ja, alle Daten vom HM und alle Daten vom Victron werden unter mqtt0 angezeigt
und jetzt hat beim Neustart der MQTT 0 Instanz automatisch die 1 sich aktualisiert und auf einmal sind unter 1 alle Daten angezeigt

**Reply by Mogdar-M** [reply]
und irgendwas ist passiert, aktuell ist das Limit auf 0 gegangen

**Reply by Mogdar-M** [reply]
und der WR hat abgeschaltet.
Also hat er zumindest einmal Daten angenommen.

Gibst du die Steuerung auf mqtt0 oder 1?

**Reply by DPO99** [reply]
Ich gebe die Steuerung immer über den MQTT.1 ein. Aber genau aus dem Grund habe ich die Einstellung entfernt, dass beim Reconnect alle States publiziert werden was zu einem nicht vorhersehbaren Ergebnis führen kann.

Sofern der WR jetzt aus ist und "ruhig" ist (sprich keine anderen Befehle querschießen) versuch doch mal den über den iobroker zu steuern.

**Comment by Mogdar-M** [comment]
ES LÄUFT!

Ich hatte unter MQTT 1 die CMD Einträge auch manuell erstellt, denn die Liste war ja leer.
Nun habe ich MQTT 1 deaktiviert, alle Einträge gelöscht, MQTT 1 wieder aktiviert und MQTT 0 neu gestartet.

Damit wurden unter MQTT 1 alle Einträge automatisch angelegt.

Und sofort geht der HM aus, denn die Werte waren alle auf NULL.
Mit diesem Blocky Script bekomme ich den HM dann limitiert:
![image](https://user-images.githubusercontent.com/129499239/235886085-c250b4ab-f99e-49ea-9e2b-096e0924a29d.png)


Danke für deine Hilfe!

**Reply by DPO99** [reply]
Super so solls sein =). Die Diskussion dann gerne schließen.

**Comment by berndroid** [comment]
Bei mir werden die Daten in mqtt.1 geschrieben und auch in mqtt.0 übernommen, aber das Limit wird immer wieder gleich mit den 1600W überschrieben, die im WR hinterlegt sind.

**Reply by DPO99** [reply]
Das klingt danach als ob du einen falschen wert oder das Objekt falsch steuerst. Es springt sicher wieder auf den alten wert zurück, da der von der dtu an den broker gesendet wird.
